pkg/activitypub/store/ariesstore/ariesstore.go

```diff
@@ -0,0 +1,453 @@
+		return nil, fmt.Errorf("failed to determine if there are more results: %w", err)
+	}
+
+	if areMoreResults {
+		activityBytes, err := a.ariesIterator.Value()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get value: %w", err)
+		}
+
+		var activity vocab.ActivityType
+
+		err = activity.UnmarshalJSON(activityBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal activity bytes: %w", err)
+		}
+
+		return &activity, nil
+	}
+
+	return nil, spi.ErrNotFound
+}
+
+func (a *activityIterator) Close() {}
+
+type referenceIterator struct {
+	ariesIterator ariesstorage.Iterator
+}
+
+// TODO (#299) return a real result instead of 0.
+func (r *referenceIterator) TotalItems() int {
+	return 0
+}
+
+func (r *referenceIterator) Next() (*url.URL, error) {
+	areMoreResults, err := r.ariesIterator.Next()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine if there are more results: %w", err)
+	}
+
+	if areMoreResults {
+		urlBytes, err := r.ariesIterator.Value()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get value: %w", err)
+		}
+
+		retrievedURL, err := url.Parse(string(urlBytes))
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse stored value as a URL: %w", err)
+		}
+
+		return retrievedURL, nil
+	}
+
+	return nil, spi.ErrNotFound
+}
+
+func (r *referenceIterator) Close() {}
+
+type stores struct {
+	activities ariesstorage.Store
+	reference  map[spi.ReferenceType]ariesstorage.Store
+	actor      ariesstorage.Store
+}
+
+func openStores(provider ariesstorage.Provider) (stores, error) {
+	activityStore, err := provider.OpenStore("activity")
+	if err != nil {
+		return stores{}, fmt.Errorf("failed to open activity store: %w", err)
+	}
+
+	err = provider.SetStoreConfig("activity",
+		ariesstorage.StoreConfiguration{
+			TagNames: []string{activityTag, timeAddedTagName},
+		})
+	if err != nil {
+		return stores{}, fmt.Errorf("failed to set store configuration on activity store: %w", err)
+	}
+
+	referenceStores, err := openReferenceStores(provider)
+	if err != nil {
+		return stores{}, fmt.Errorf("failed to open reference stores: %w", err)
+	}
+
+	actorStore, err := provider.OpenStore("actor")
+	if err != nil {
+		return stores{}, fmt.Errorf("failed to open actor store: %w", err)
+	}
+
+	return stores{
+		activities: activityStore,
+		reference:  referenceStores,
+		actor:      actorStore,
+	}, nil
+}
+
+func openReferenceStores(provider ariesstorage.Provider) (map[spi.ReferenceType]ariesstorage.Store, error) {
+	referenceTypes := []spi.ReferenceType{
+		spi.Inbox, spi.Outbox, spi.Follower, spi.Following, spi.Witness,
+		spi.Witnessing, spi.Like, spi.Liked, spi.Share, spi.AnchorCredential,
+	}
+
+	storeConfig := ariesstorage.StoreConfiguration{
+		TagNames: []string{objectIRITagName, timeAddedTagName},
+	}
+
+	referenceStores := make(map[spi.ReferenceType]ariesstorage.Store)
+
+	for _, referenceType := range referenceTypes {
+		store, err := provider.OpenStore(string(referenceType))
```

Comment:
Is there any overhead to having this many stores? If so, can the reference type be handled using tags?

Reply:
Hmm... this is a tricky thing to figure out. It is possible to actually store everything all in one store (which = one database in CouchDB), and just use tags to grab the data you want. The thing I wonder is this: is it more efficient to spread out the data among databases in CouchDB? If everything is all in one big database, then the same indexes will get updated more frequently... and I wonder if somehow this may be less efficient than having multiple smaller databases (each having smaller indices as a result as well). Perhaps it scales better in a cluster when you have multiple databases this way? Not sure. I guess I would need to do some testing to figure this out (although it wouldn't be too easy to do this test right now). What do you think?

Reply:
For now we can keep it the way you have it.